vm/starnix/starnix.go

```diff
@@ -0,0 +1,374 @@
+// Copyright 2023 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package starnix
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/google/syzkaller/pkg/config"
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/pkg/osutil"
+	"github.com/google/syzkaller/pkg/report"
+	"github.com/google/syzkaller/sys/targets"
+	"github.com/google/syzkaller/vm/vmimpl"
+)
+
+func init() {
+	var _ vmimpl.Infoer = (*instance)(nil)
+	vmimpl.Register("starnix", ctor, true)
+}
+
+type Config struct {
+	// Number of VMs to run in parallel (1 by default).
+	Count int `json:"count"`
+	// Location of Fuchsia checkout, for now we need this to run ffx commands.
+	Fuchsia string `json:"fuchsia"`
+}
+
+type Pool struct {
+	count   int
+	env     *vmimpl.Env
+	cfg     *Config
+	version string
+}
+
+type instance struct {
+	fuchsiaDirectory string
+	name             string
+	index            int
+	cfg              *Config
+	version          string
+	args             []string
+	debug            bool
+	workdir          string
+	timeouts         targets.Timeouts
+	port             int
+	rpipe            io.ReadCloser
+	wpipe            io.WriteCloser
+	fuchsiaVmStarted bool
+	fuchsiaLogs      *exec.Cmd
+	adb              *exec.Cmd
+	executor         string
+	merger           *vmimpl.OutputMerger
+	diagnose         chan bool
+}
+
+func getFuchsiaVersion(fuchsiaDirectory string) ([]byte, error) {
+	cmd := osutil.Command("ffx", "version")
+	cmd.Dir = fuchsiaDirectory
+	return cmd.Output()
+}
+
+func ctor(env *vmimpl.Env) (vmimpl.Pool, error) {
+	cfg := &Config{}
+	if err := config.LoadData(env.Config, cfg); err != nil {
+		return nil, fmt.Errorf("failed to parse starnix vm config: %v", err)
+	}
+	if cfg.Count < 1 || cfg.Count > 128 {
+		return nil, fmt.Errorf("invalid config param count: %v, want [1, 128]", cfg.Count)
+	}
+	if !osutil.IsDir(cfg.Fuchsia) {
+		return nil, fmt.Errorf("%q is not a directory", cfg.Fuchsia)
+	}
+	version, err := getFuchsiaVersion(cfg.Fuchsia)
+	if err != nil {
+		return nil, fmt.Errorf("There is an error running ffx commands in the Fuchsia checkout (%q): %v", cfg.Fuchsia, err)
+	}
+
+	pool := &Pool{
+		count:   cfg.Count,
+		env:     env,
+		cfg:     cfg,
+		version: string(version),
+	}
+	return pool, nil
+}
+
+func (pool *Pool) Count() int {
+	return pool.count
+}
+
+func (pool *Pool) Create(workdir string, index int) (vmimpl.Instance, error) {
+	return pool.ctor(workdir, index)
```

Comment:
Move ctor function contents here. This indirection only adds code and complexity.